device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/MultiplexingClient.java

```diff
@@ -0,0 +1,232 @@
+     * <p>
+     * The registered device client instance must use the same transport protocol (AMQPS or AMQPS_WS) that this multiplexing client uses.
+     * <p>
+     * The registered device client may have its own retry policy and its own SAS token expiry time, separate from every other multiplexed device.
+     * <p>
+     * The registered device client must use symmetric key based authentication.
+     * <p>
+     * The registered device client must belong to the same IoT Hub as all previously registered device clients.
+     * <p>
+     * If this device client is already registered, then this method will throw a {@link IllegalStateException}
+     * <p>
+     * @param deviceClient The device client to associate with this multiplexing client.
+     */
+    public void registerDeviceClient(DeviceClient deviceClient)
+    {
+        Objects.requireNonNull(deviceClient);
+
+        for (DeviceClient currentClient : deviceClientList)
+        {
+            if (currentClient.getConfig().getDeviceId().equals(deviceClient.getConfig().getDeviceId()))
+            {
+                throw new IllegalStateException("Cannot register device client because a device client for that device identity has already been registered");
+            }
+        }
+
+        if (deviceClient.getConfig().getAuthenticationType() != DeviceClientConfig.AuthType.SAS_TOKEN)
+        {
+            throw new UnsupportedOperationException("Can only register a device client if it uses SAS token based authentication");
+        }
+
+        if (deviceClient.getConfig().getProtocol() != this.protocol)
+        {
+            throw new UnsupportedOperationException("Cannot register a device client that uses a different transport protocol than this multiplexing client uses");
+        }
+
+        if (this.protocol == IotHubClientProtocol.AMQPS && this.deviceClientList.size() > 1000)
+        {
+            throw new UnsupportedOperationException("Multiplexed connections over AMQPS only support up to 1000 devices");
+        }
+
+        // Typically client side validation is duplicate work, but IoT Hub doesn't give a good error message when closing the
+        // AMQPS_WS connection so this is the only way that users will know about this limit
+        if (this.protocol == IotHubClientProtocol.AMQPS_WS && this.deviceClientList.size() > 500)
+        {
+            throw new UnsupportedOperationException("Multiplexed connections over AMQPS_WS only support up to 500 devices");
+        }
+
+        if (this.deviceClientList.size() > 1 && !this.deviceClientList.get(0).getConfig().getIotHubHostname().equals(deviceClient.getConfig().getIotHubHostname()))
+        {
+            throw new UnsupportedOperationException("Multiplexed device clients must connect to the same host name");
+        }
+
+        // The first device to be registered will cause this client to build the IO layer with its configuration
+        if (this.deviceIO == null)
```

Comment:
This doesn't look threadsafe. Shouldn't it be?

Reply:
It should be. I'll add a lock to synchronize the register/unregister calls onto

Reply:
I think 186 - 200 should be in a synchronized block